packages/compartment-mapper/test/make-import-hook-maker.test.js

```diff
@@ -0,0 +1,156 @@
+/* eslint-disable no-shadow */
+import './ses-lockdown.js';
+import test from 'ava';
+import { mapNodeModules } from '../src/node-modules.js';
+import { makeProjectFixtureReadPowers } from './project-fixture.js';
+import { defaultParserForLanguage } from '../src/import-parsers.js';
+import { captureFromMap } from '../capture-lite.js';
+import { loadFromMap } from '../import-lite.js';
+import { loadLocation } from '../import.js';
+
+/**
+ * @import {
+ *   CanonicalName,
+ *   MaybeReadPowers,
+ *   ModuleSourceHook,
+ * } from '../src/types.js';
+ * @import {ExecutionContext} from 'ava';
+ * @import {ProjectFixture} from './test.types.js';
+ */
+
+/**
+ * Test fixture with a simple entry module
+ * @type {ProjectFixture}
+ */
+const testFixture = {
+  root: 'test-app',
+  graph: {
+    'test-app': [],
+  },
+  entrypoint: 'file:///node_modules/test-app/index.js',
+};
+
+/**
+ * AVA macro for testing moduleSource hook with different loader functions
+ *
+ * This could be more narrowly-typed, but it's a test.
+ *
+ * @param {ExecutionContext} t
+ * @param {object} options
+ * @param {Function} options.loaderFn - The function to call (captureFromMap, loadFromMap, or loadLocation)
+ * @param {(readPowers: MaybeReadPowers, entrypoint: string, options: object) => any} options.argsFn - Function that returns arguments for the loader function
+ * @param {boolean} options.callImport - Whether to call application.import() after loading
+ */
+const moduleSourceHookTest = async (
+  t,
+  { loaderFn, argsFn, callImport = false },
+) => {
+  t.plan(7); // log, moduleSource, canonicalName type, hook called, canonicalName content, moduleSource type, moduleSource truthy
+
+  const readPowers = makeProjectFixtureReadPowers(testFixture);
+
+  let hookCalled = false;
+  /** @type {object | undefined} */
+  let receivedModuleSource;
+  /** @type {CanonicalName | undefined} */
+  let receivedCanonicalName;
+
+  /** @type {ModuleSourceHook} */
+  const moduleSourceHook = ({ moduleSource, canonicalName, log }) => {
+    hookCalled = true;
+    receivedModuleSource = moduleSource;
+    receivedCanonicalName = canonicalName;
+
+    t.truthy(log, 'log function should be provided');
+    t.truthy(moduleSource, 'moduleSource should be provided');
+    t.is(typeof canonicalName, 'string', 'canonicalName should be a string');
+  };
+
+  const options = {
+    moduleSourceHook,
+    parserForLanguage: defaultParserForLanguage,
+  };
+
+  const args = await argsFn(
+    readPowers,
+    /** @type {string} */ (testFixture.entrypoint),
+    options,
+  );
```

Comment:
(nit) feels like having a single function `doTheLoading` that takes the argsFn arguments and puts things together in one pass would be easier to follow. And the testcase code would resembl the expected usage pattern of the `mapNodeModules` and `captureFromMap` more closely.